Changed procedure annotation

OliviaYtterbrink opened the pull request:
New annotation takes mode as a parameter, which can be one of `READ`, `WRITE`, `SCHEMA`, or `DBMS`.
Deprecates `@PerformsWrites`

changelog: Added mode to Procedure annotation, deprecates PerformsWrites

OliviaYtterbrink committed:
Extended Procedure annotation

- @Procedure now accepts extra parameters for scope and access level
Only scope = DBMS and access = WRITE is currently supported

OliviaYtterbrink committed:
Implemented Schema scope for procedures

- Specifying scope = SCHEMA gives the procedure permission to write to
 the schema. This can't be combined with writing to the graph.

OliviaYtterbrink committed:
Cleaned up formatting

OliviaYtterbrink committed:
Merged parameters of procedure

Scope and Access has been merged to Mode

OliviaYtterbrink committed:
Updated documentation

OliviaYtterbrink committed:
Deprecated @PerformsWrites

jexp commented:
yay! Thanks a lot!

lutovich commented:
This is a really nice change! I still think though that `Procedure.Mode = DatabaseManagementSystem` does not make much sense. Maybe we should use `DATABASE_MANAGEMENT` or `USER_MANAGEMENT` instead of `DBMS`.

davidegrohmann commented:
What happens if someone mix the old annotation with the new ones in an inconsistent way? E.g, something like:

```
 @PerformsWrites
 @Procedure( mode = READ )
```

Can we add tests that ensure that we fail in such cases and we do not accept the procedure?

OliviaYtterbrink committed:
Added handling of conflicting modes

- Annotating with @PerformsWrites is only accepted with mode READ or WRITE.
  With SCHEMA and DBMS an error will be thrown.

OliviaYtterbrink commented:
Good catch, added tests.
`@PerformsWrites` with `mode = READ` or `mode = WRITE` results in a write procedure.
If it is combined with the other modes, it will generate an error.

davidegrohmann commented:
@OliviaYtterbrink isn't it better to fail in the case `@PerformsWrites` and `mode = READ` instead trying to guess what the user wanted?

fickludd commented:
@davidegrohmann yes, but we run into problems with backwards compatibility.
`@PerformsWrites @Procedure`
Should be a valid annotation, and this would yield such a collision. On the plus side, mode = READ is a subset of mode = WRITE, so the conflict is not so bad. 

To solve the explicit conflict, we could create a special value `DEFAULT_READ` that could be undocumented, set as default, and be the only allowable combination with `@PerformsWrites`. Then we could make `@PerformsWrites @Procedure( mode = "READ" )` fail.

davidegrohmann commented:
@fickludd The conflict might be very bad if the user intended the procedures to be read only and forgot to delete the old `PerformsWrites` annotation.

Moreover I am not following what the problem is, the annotation `@Procedure` is old by the param `mode` is brand new, so it should be possible to fail in case of that conflict. No?

henriknyman commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -1103,6 +1103,12 @@ public void uniqueIndexDrop( IndexDescriptor descriptor ) throws DropIndexFailur
+    @Override
+    public RawIterator<Object[], ProcedureException> procedureCallSchema( ProcedureName name, Object[] input ) throws ProcedureException
+    {
+        return callProcedure( name, input, AccessMode.Static.FULL );
```
Maybe now that we have procedureCallSchema we can lower the access mode of procedureCallWrite from FULL to WRITE (see the TODO in this file)

henriknyman commented:
We will go with what @davidegrohmann suggested and fail in the case of mixing old and new annotations, but it will be added on by this separate PR: https://github.com/neo4j/neo4j/pull/7488
